File: implementations/java-springboot/examples/src/test/java/com/example/accountservice/card/application/command/CancelCardsByAccountServiceTest.java
package com.example.accountservice.card.application.command;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.accountservice.card.domain.Card;
import com.example.accountservice.card.domain.CardFindQuery;
import com.example.accountservice.card.domain.CardRepository;
import com.example.accountservice.card.domain.CardStatus;
import com.example.accountservice.card.domain.CardsWithCount;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class CancelCardsByAccountServiceTest {

    @Mock private CardRepository cardRepository;

    private CancelCardsByAccountService service;

    @BeforeEach
    void setUp() {
        service = new CancelCardsByAccountService(cardRepository);
    }

    @Test
    void cancels_and_saves_all_of_the_accounts_ACTIVE_and_SUSPENDED_cards() {
        Card active = Card.issue("account-1", "owner-1", "VISA");
        Card suspended = Card.issue("account-1", "owner-1", "MASTER");
        suspended.suspend();
        when(cardRepository.findCards(
                        new CardFindQuery(
                                0,
                                Integer.MAX_VALUE,
                                null,
                                null,
                                "account-1",
                                List.of(CardStatus.ACTIVE, CardStatus.SUSPENDED))))
                .thenReturn(new CardsWithCount(List.of(active, suspended), 2));

        service.cancel(new CancelCardsByAccountCommand("account-1"));

        verify(cardRepository, times(2)).saveCard(org.mockito.ArgumentMatchers.any());
    }

    @Test
    void saves_nothing_when_there_are_no_cards_to_cancel_idempotency() {
        when(cardRepository.findCards(
                        eq(
                                new CardFindQuery(
                                        0,
                                        Integer.MAX_VALUE,
                                        null,
                                        null,
                                        "account-1",
                                        List.of(CardStatus.ACTIVE, CardStatus.SUSPENDED)))))
                .thenReturn(new CardsWithCount(List.of(), 0));

        service.cancel(new CancelCardsByAccountCommand("account-1"));

        verify(cardRepository, never()).saveCard(org.mockito.ArgumentMatchers.any());
    }
}

File: implementations/kotlin-springboot/examples/src/test/kotlin/com/example/accountservice/card/application/command/CancelCardsByAccountServiceTest.kt
package com.example.accountservice.card.application.command

import com.example.accountservice.card.domain.Card
import com.example.accountservice.card.domain.CardFindQuery
import com.example.accountservice.card.domain.CardRepository
import com.example.accountservice.card.domain.CardStatus
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class CancelCardsByAccountServiceTest {
    private val cardRepository = mockk<CardRepository>(relaxed = true)
    private val service = CancelCardsByAccountService(cardRepository)

    private fun card(
        cardId: String,
        status: CardStatus,
    ): Card =
        Card.reconstitute(
            cardId = cardId,
            accountId = "account-1",
            ownerId = "owner-1",
            brand = "VISA",
            status = status,
            createdAt = LocalDateTime.now(),
        )

    private val findQuery =
        CardFindQuery(
            page = 0,
            take = 1000,
            accountId = "account-1",
            status = listOf(CardStatus.ACTIVE, CardStatus.SUSPENDED),
        )

    @Test
    fun `cancels and saves all ACTIVE and SUSPENDED cards for the account`() {
        val cards = listOf(card("card-1", CardStatus.ACTIVE), card("card-2", CardStatus.SUSPENDED))
        every { cardRepository.findCards(findQuery) } returns (cards to cards.size.toLong())

        service.cancel("account-1")

        cards.forEach { assertThat(it.status).isEqualTo(CardStatus.CANCELLED) }
        verify(exactly = 2) { cardRepository.saveCard(any()) }
    }

    @Test
    fun `saves nothing when there are no cards to cancel (idempotent)`() {
        every { cardRepository.findCards(findQuery) } returns (emptyList<Card>() to 0L)

        service.cancel("account-1")

        verify(exactly = 0) { cardRepository.saveCard(any()) }
    }
}
